pkg/querier/querier.go

```diff
@@ -191,28 +202,86 @@ func (q querier) Select(sp *storage.SelectParams, matchers ...*labels.Matcher) (
 		}
 	}
 
-	return storage.NewMergeSeriesSet(result, nil), nil, nil
+	// we have all the sets from different sources (chunk from store, chunks from ingesters,
+	// time series from store and time series from ingesters).
+	return q.mergeSeriesSets(result), nil, nil
 }
 
 // LabelsValue implements storage.Querier.
 func (q querier) LabelValues(name string) ([]string, storage.Warnings, error) {
-	lv, err := q.distributor.LabelValuesForLabelName(q.ctx, model.LabelName(name))
-	return lv, nil, err
+	return q.primary.LabelValues(name)
 }
 
 func (q querier) LabelNames() ([]string, storage.Warnings, error) {
-	ln, err := q.distributor.LabelNames(q.ctx)
-	return ln, nil, err
+	return q.primary.LabelNames()
 }
 
-func (q querier) metadataQuery(matchers ...*labels.Matcher) (storage.SeriesSet, storage.Warnings, error) {
-	ms, err := q.distributor.MetricsForLabelMatchers(q.ctx, model.Time(q.mint), model.Time(q.maxt), matchers...)
-	if err != nil {
-		return nil, nil, err
+func (querier) Close() error {
+	return nil
+}
+
+func (q querier) mergeSeriesSets(sets []storage.SeriesSet) storage.SeriesSet {
```

Comment:
Is this unit tested? I can't see its tests, but maybe I'm missing something.

Reply:
Yes, TestQuerier function exercises all of this functionality. Merging multiple chunks is tested when streaming is on, while merging chunks and non-chunk-based sets happens with streaming off.